lib/java/src/org/apache/thrift/transport/TFastFramedTransport.java

```diff
@@ -172,11 +177,12 @@ public void clear() {
 
   @Override
   public void flush() throws TTransportException {
-    int length = writeBuffer.getPos();
-    TFramedTransport.encodeFrameSize(length, i32buf);
-    underlying.write(i32buf, 0, 4);
-    underlying.write(writeBuffer.getBuf().array(), 0, length);
+    int length = writeBuffer.getPos() - 4;          // account for frame size reserved space
+    byte[] data = writeBuffer.getBuf().array();
+    TFramedTransport.encodeFrameSize(length, data); // write frame size directly into the buffer
+    underlying.write(data, 0, length + 4);          // write the frame size and payload in one call
     writeBuffer.reset();
+    writeBuffer.write(sizeFiller, 0, 4);            // reserve space for the next frame size
```

Comment:
I understand the optimization, but worry that now there are two places adding `sizeFiller` to the buffer.

Reply:
Agreed; reset() could do this if it were in the writeBuffer instead.